src/main/scala/com/microsoft/hyperspace/actions/OptimizeAction.scala

```diff
@@ -106,20 +105,39 @@ class OptimizeAction(
 
     if (filesToOptimize.isEmpty) {
       throw NoChangesException(
-        "Optimize aborted as no index files smaller than " +
+        "Optimize aborted as no applicable files smaller than " +
           s"${HyperspaceConf.optimizeFileSizeThreshold(spark)} found.")
     }
   }
 
   private lazy val (filesToOptimize, filesToIgnore): (Seq[FileInfo], Seq[FileInfo]) = {
-    if (mode.equalsIgnoreCase(IndexConstants.OPTIMIZE_MODE_QUICK)) {
-      val threshold = HyperspaceConf.optimizeFileSizeThreshold(spark)
-      previousIndexLogEntry.content.fileInfos.toSeq.partition(_.size < threshold)
-    } else {
-      // For 'full' mode, put all the existing index files into 'filesToOptimize' partition so
-      // that one file is created per bucket.
-      (previousIndexLogEntry.content.fileInfos.toSeq, Seq())
+    val (possibleCandidates, largeFilesToIgnore) =
+      if (mode.equalsIgnoreCase(IndexConstants.OPTIMIZE_MODE_QUICK)) {
+        val threshold = HyperspaceConf.optimizeFileSizeThreshold(spark)
+        previousIndexLogEntry.content.fileInfos.toSeq.partition(_.size < threshold)
+      } else {
+        // For 'full' mode, put all the existing index files into 'filesToOptimize' partition
+        // so that one file is created per bucket.
+        (previousIndexLogEntry.content.fileInfos.toSeq, Seq())
+      }
+
+    // Check if there is only one index file in a bucket and exclude the index file
+    // as it doesn't need to be optimized.
+    val filesPerBucket = mutable.Map[String, List[FileInfo]]()
+    possibleCandidates.foreach { f =>
+      // Extract key from the file path; "part-XXXXX" indicates each bucket id.
+      val key = new Path(f.name).getName.substring(0, 10)
```

Comment:
I think we should use Spark's `getBucketId()` function to extract bucket-id from a FileName. 
Take a look at below: 
https://github.com/apache/spark/blob/5264164a67df498b73facae207eda12ee133be7d/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/BucketingUtils.scala#L36

Reply:
Thanks!